File: Graphs/Depth-First-Search, Topological Sorting/DFS.py
# DFS Implementation using Adjacency list
# Author - rudrajit1729

from collections import defaultdict

# Class representing graph as adjacency list
class Graph:
	def __init__ (self, V):
		self.graph = defaultdict(list)
		self.V = V # Number of vertices

	def addEdge(self, u, v): # Adds an edge in the graph
		self.graph[u].append(v)

	def DFS_Visit(self, s, visited): # DFS for a node
		visited[s] = True
		print(s, end = " ")

		for v in self.graph[s]:
			if not visited[v]:
				self.DFS_Visit(v, visited)

	def DFS(self): # DFS traversal of graph
		visited = [False] * self.V

		for s in range(self.V):
			if not visited[s]:
				self.DFS_Visit(s, visited)


def main():
	g = Graph(6)
	g.addEdge(5, 0)
	g.addEdge(5, 2)
	g.addEdge(4, 0)
	g.addEdge(4, 1)
	g.addEdge(2, 3)
	g.addEdge(3, 1)
	print("DFS traversal of the graph G: ")
	g.DFS()

	f = Graph(4)
	f.addEdge(0, 1) 
	f.addEdge(0, 2) 
	f.addEdge(1, 2) 
	f.addEdge(2, 0) 
	f.addEdge(2, 3) 
	f.addEdge(3, 3) 
	print("\nDFS traversal of the graph F: ")
	f.DFS()

if __name__ == '__main__':
	main()

File: Graphs/Depth-First-Search, Topological Sorting/DFS.cpp
// DFS Implementation using Adjacency list
// Author - rudrajit1729

#include<bits/stdc++.h>
using namespace std;

class Graph
{
	list<int> *graph;
	int V;
public:
	Graph(int V);
	void addEdge(int u, int v);
	void DFS_Visit(int s, bool *visited);
	void DFS();
};

Graph::Graph(int V)
{
	this->V = V;
	graph = new list<int> [V];
}

void Graph::addEdge(int u, int v)
{
	graph[u].push_back(v);
}

void Graph::DFS_Visit(int s, bool *visited)
{
	cout<<s<<"\t";
	visited[s] = true;
	for(int v : graph[s])
	{
		if (!visited[v])
			{
				DFS_Visit(v, visited);
			}
	}	
}

void Graph::DFS()
{
	bool *visited = new bool[V];
	for(int i = 0; i< V; i++) visited[i] = false;
	for(int s = 0; s < V; s++)
	{
		if (!visited[s])
		{
			DFS_Visit(s, visited);
		}
	}
}

int main() 
{ 
    // Create a graph given in the above diagram 
    // Graph g(4); 
    // g.addEdge(0, 1); 
    // g.addEdge(0, 2); 
    // g.addEdge(1, 2); 
    // g.addEdge(2, 0); 
    // g.addEdge(2, 3); 
    // g.addEdge(3, 3); 
  	Graph g(6);
	g.addEdge(5, 0);
	g.addEdge(5, 2);
	g.addEdge(4, 0);
	g.addEdge(4, 1);
	g.addEdge(2, 3);
	g.addEdge(3, 1);
    cout << "Following is Depth First Traversal \n"; 
    g.DFS(); 
  	cout<<endl;
    return 0; 
} 